### PR TITLE
Add a download set.

### PR DESCRIPTION
**Pull request by hdevalence**
WIP, cannot be tested because state changes broke the sync component.  Based on top of #1040.

### CODE REVIEW

**Comment by hdevalence** [comment]
Rebased onto main to pick up the state fixes and the improved message handling work.

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs]: Consider using ? rather than return - it provides better error context.
```suggestion
                Err(eyre!("queued download of hash behind our chain tip"))?;
```

**Review by teor2345 — COMMENTED** [review]
This is a good refactor - it should make the sync process easier to understand and maintain.

I'm looking forward to seeing the test results. Let me know when you're ready for a review or testing.

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs]: How is this error handled or mitigated by the changes in this PR?

As far as I understand, the cancellation in `Downloads` mitigates "duplicate block" errors. But I can't see any new code in this PR that handles or mitigates ExtendTips errors.

So I wonder if we'll still need this retry, particularly on testnet.

Comment (hdevalence) [reply zebrad/src/components/sync.rs]: If we fail to extend the tip set, then we run out of prospective tips and restart the sync procedure.  This PR tries to make restarts cheap, so that restarting doesn't cause a cascading failure.

**Comment by hdevalence** [comment]
Rebased onto main again and redid the cancellation handling in the `Connection` state machine to cancel pending requests (i.e., handle request cancellations that happen after the request arrives at the connection).

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs]: Hmm, adding this change made clippy unhappy.

Comment (yaahc) [reply zebrad/src/components/sync.rs]: yes, it do that, ignore clippy on this one, it's wrong. I have been meaning to disable that lint for ages with the intent of eventually introducing the opposite lint.

Comment (teor2345) [reply zebrad/src/components/sync.rs]: Weird, I thought we'd added allow(try_err) a few weeks ago

Comment (teor2345) [reply zebrad/src/components/sync.rs]: Hmm, apparently not, it's not in lib.rs.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs]: Unresolved since I'm not sure this approach actually works on testnet.

Comment (teor2345) [reply zebrad/src/components/sync.rs]: Yeah I wonder if we should just deploy a few Zebra or zcashd nodes to stabilise testnet.

Or if we should spend the effort to make sync work under difficult network conditions.

File: zebrad/src/components/sync/downloads.rs
Comment (dconnolly) [inline zebrad/src/components/sync/downloads.rs:125]: 👍

**Review by dconnolly — COMMENTED** [review]
Approved once the clippy error is fixed

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs]: Can we split the ExtendTips change into another PR, and merge the rest of this PR?

**Comment by teor2345** [comment]
Could this PR fix the sync hang in #1181?

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs]: @hdevalence can we split the ExtendTips change into another PR?

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs]: Let's add `#![allow(clippy::try_err)]` to `zebrad/src/lib.rs` ?

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs]: No, I'd rather fix it here.

**Review by teor2345 — CHANGES_REQUESTED** [review]
The code looks good, but I haven't really had a chance to run it.

This PR doesn't contain any tests, what's the plan for testing it?

**Comment by teor2345** [comment]
PR #1193 adds a limited sync test for the first large checkpoint, but it won't find later stalls or errors.

File: zebrad/src/components/sync.rs
Comment (dconnolly) [inline zebrad/src/components/sync.rs]: 👍

File: zebrad/src/components/sync.rs
Comment (dconnolly) [inline zebrad/src/components/sync.rs]: cancelled

**Review by dconnolly — DISMISSED** [review]
Looks good to me but I'll wait on merging per others' feedback

**Comment by hdevalence** [comment]
This includes the changes from #1089, in an effort to get the simplified sync logic done in one change.  The sync algorithm as designed had some invariants that the different parts were supposed to maintain together, and I think we ran into problems changing some of the parts one at a time.  So, this PR tries to clean up the sync algorithm in a coherent way.  I tested it on mainnet and testnet.  On testnet it makes progress but is very slow, on account of the bad condition of the testnet.

There are some ancillary changes made along the way while testing, mainly to the span and tracing handling. There's a new debug line added before queuing downloads that reveals that a big contributor to latency of the sync while checkpointing mainnet is the state service (checking that new hashes are not already part of the state).  I think that this could be safely omitted from the extendtips step, but not obtaintips.  (That was part of the original motivation for having those be two separate steps).

File: zebrad/src/components/sync.rs
Comment (hdevalence) [inline zebrad/src/components/sync.rs:82]: @teor2345 I turned this down to accelerate the restart process.  It seems to still work on testnet.  Although it gets stuck sometimes, it does eventually make progress.  We can turn it back up if problems re-occur, but it seemed like faster restarts on mainnet were more important.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [inline zebrad/src/components/sync.rs:199]: Matching on the error value lets us display the error to the user.  We don't need to check whether prospective_tips is empty, because we'll just fall through to the end of the loop.

In a later PR we should try to change the errors so that they don't display a bug report link -- these errors are likely network errors or transient errors (cc @yaahc)

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs]: Let's wait until there's a lint one way or the other, then use it to ensure consistency.

Comment (teor2345) [reply zebrad/src/components/sync.rs]: I'm confused. Do you think the existing clippy lint is good advice?
Or do you think that consistency with all the default linter rules is really important?

I suggested that we use the ? operator, because it provides better error context.

I believe that benefit outweighs the cost of any minor inconsistencies that might happen from disabling a single lint.

(If we wait for a changed lint to hit clippy stable, we might be waiting months. In the meantime, our error reporting will be worse.)

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:82]: I'd really like to reduce the sync timeout, but I'd also like to make sure it works across a range of network conditions.

Have you tested it from a low bandwidth or high latency network location?

If not, I can do those tests tomorrow.

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs]: I agree that it's a bad idea to have the lint on for some crates, and off for others.

I can submit a PR that turns it off everywhere - it's already disabled in most of the large zebra crates.

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:199]: (Let's open a ticket, so we don't forget to make this change.)

File: zebra-consensus/src/checkpoint.rs
Comment (teor2345) [inline zebra-consensus/src/checkpoint.rs:509]: I find this log really useful for diagnostics, can we keep it?

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs:36]: It would be nice to get rid of the outer Retry, but I think that would mean adding a retry limit within the Hedge.

And maybe we'd also want to set a mandatory number of retries, even when there are too few data points in the histogram.

Comment (teor2345) [reply zebrad/src/components/sync.rs:36]: But that's nowhere near what Hedge can do now, so we should keep the outer Retry.

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:82]: In particular, I'm concerned about network-limited nodes and download floods.

But hopefully DownloadSet reduces duplicates, and Hedge reduces failures,

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs]: I'd like to reduce the number of state calls. Here are two possible optimisations:

1. Check the state's tip height once per request_blocks. Reject any blocks that have a coinbase height that is lower than the tip height minus the reorg limit. They can't possibly be valid, because we already have a finalised block for that height.

2. Check if the download set has queued the hash - if it has, ignore (or reject?) the request. Then check the state, and restart if the hash is in the state. If the block is in the download set, it can't possibly be in the state, because a previous state check would have caught it.

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs:156]: I think the Retry should be outside the Hedge:
* with Hedge outside Retry, a slow request with a quick first retry gets tried 3 times: one for the initial slow request, once for the Retry on that request, and once by the quick Hedge retry
* with Retry outside Hedge, a slow request with a quick first retry gets tried 2 times: once for the initial slow request, and once by the quick Hedge retry

I think that's the way it works - but I might be missing some subtlety about the concurrency, selects and timeouts here.

If that's the case, let's explain why Hedge is outside Retry in the comments.

Comment (hdevalence) [reply zebrad/src/components/sync.rs:156]: The Hedge goes outside the Retry, because the Retry layer abstracts away spurious failures from individual peers making a less-fallible network service, and the Hedge layer tries to reduce latency of that less-fallible service.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs]: I'm not sure either of those optimizations work — in case (1), we don't have access to heights, only hashes, so I'm not sure we can do that query, and in case (2), we almost never duplicate download requests, so we wouldn't save much.

I think a better solution would be along the lines of the one I mentioned in my comment on this rebase, to eliminate state checks from extendtips invocations.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs:199]: Yes, I just wanted to cc @yaahc here to flag it in context and have some preliminary discussion first.

Comment (yaahc) [reply zebrad/src/components/sync.rs:199]: oh interesting, so we want to be able to mark _all_ errors that pass thru this point so that they don't have bug report urls presented with them. that shouldn't be a difficult feature to add. At this point tho there are a bunch of features in color-eyre being motivated by zebra and some other integrations that I've wanted to write that I couldn't easily express so I think we might want to write our own EyreHandler, I'm thinking in a zebra-eyre crate, that we can use to more tightly control these integrations and provide a staging ground where we can test out ideas before I upstream them into color-eyre.

Comment (hdevalence) [reply zebrad/src/components/sync.rs:199]: I'm not sure what the easiest way to implement it would be, but if there were a method that could turn on or off the bugreport URL for a specific error instance, I think it would be fine to go through and tack it on to every `tracing::warn!` invocation in the loop.

Having a `zebra-eyre` crate could be good.  My only concern is that once that crate exists, it becomes an easy "accumulation point" for extra code, and in the long run I think it would be good if we didn't have any zebra-specific error handling.  So it would be good to have a plan or focus on how to ensure that the staging ground remains a staging ground.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs:36]: I don't follow -- hedged requests send up to one additional request, what would a retry limit within the Hedge look like?

File: zebra-consensus/src/checkpoint.rs
Comment (hdevalence) [reply zebra-consensus/src/checkpoint.rs:509]: My preference would be to remove it, since the checkpoint verifier is working now, and download progress is better monitored using the metrics integration.

File: zebrad/src/components/sync.rs
Comment (yaahc) [reply zebrad/src/components/sync.rs:199]: `zebra-eyre` would be specific to error reporting, not error handling, so I think it's fine to have specific rules. The default `std::error::Error` integration will make it possible to get good reporting for errors regardless for anyone using our libraries, and if they want to use `eyre` they could then also install our `zebra-eyre` error reporting hook to get customized reporting that is more aware of the concrete types we use and is able to extract more information that isn't exposed through the current version of the error trait.

This will make it a lot easier for us to capture and expose `SpanTrace`s in our libraries that can then be printed in our error reports, which will make our error messages much more informative I feel.

**Comment by hdevalence** [comment]
This code has some spurious error messages on slow networks, caused by https://github.com/tower-rs/tower/issues/475.  These errors are functionally unimportant and can be fixed upstream.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs]: I don't think that it's particularly important either way, and in this case I think that it makes no meaningful difference at all. I'd be happy using one, or the other. But I would prefer that if we want to make a policy choice, we do that in a consistent way, independent of this PR.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs:199]: Cool, that sounds fine.  Would you like to make an issue with that plan?

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:36]: Instead of having an outer retry and an inner hedge with 1 additional request, the hedge could send up to `N` additional requests.

Maybe that's actually more complicated though - because you have to decide when to send the other requests.

This is just speculation - it's not that important.

File: zebra-consensus/src/checkpoint.rs
Comment (teor2345) [reply zebra-consensus/src/checkpoint.rs:509]: The metrics integration works for some users, but others won't necessarily have it set up.

If we don't log anything, we'll get bug reports which say that Zebra isn't making any progress.

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs]: Ok, this is now ticket #1199.

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:82]: Our plan is to:
- [ ] Test this change on a slower network using mainnet
- [ ] Test this change on testnet, but treat stalls as a lower priority, because testnet doesn't have many nodes
- [ ] Deploy some extra `zcashd` nodes on testnet, then re-test this change on testnet

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs]: Ok, sounds good.

Have you made those changes to ExtendTips yet?
Or are they in another ticket?

Comment (teor2345) [reply zebrad/src/components/sync.rs]: Ah, I just saw 5477ce141580f108ff711bdf0913886b077b35db

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs:145]:
```suggestion
        // layer is relatively unimportant, because slow requests will
        // probably be pre-emptively hedged.
        //
        // The Hedge goes outside the Retry, because the Retry layer
        // abstracts away spurious failures from individual peers
        // making a less-fallible network service, and the Hedge layer
        // tries to reduce latency of that less-fallible service.
        //
```

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:156]: I added this comment in a suggestion.

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs:184]: 👍

File: zebra-consensus/src/checkpoint.rs
Comment (teor2345) [reply zebra-consensus/src/checkpoint.rs:509]: Ah, right, but we're still getting a log for each verified checkpoint, so this isn't a big issue.

**Review by teor2345 — COMMENTED** [review]
Looks good so far!

I'm going to test on mainnet, and see what happens on testnet, but ignore any stalls.

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:82]: zebrad mainnet stall

I'm seeing frequent stalls on mainnet, which briefly recover, then stall again.

I suspect that the sync parameters are over-tuned for high-bandwidth, low-latency, reliable networks.

Comment (teor2345) [reply zebrad/src/components/sync.rs:82]: This might actually be an issue with the block verifier or state service, I'll retry with an empty state.
Block Verifier

Comment (teor2345) [reply zebrad/src/components/sync.rs:82]: Yep, 420.3K is the sapling checkpoint height, so it's a state or block verifier issue.

File: zebrad/src/components/sync.rs
Comment (teor2345) [inline zebrad/src/components/sync.rs:54]: Since Hedge doesn't support timeouts 10 seconds or greater, should we reduce the block download timeout?
```suggestion
const BLOCK_DOWNLOAD_TIMEOUT: Duration = Duration::from_secs(9);
```

**Review by teor2345 — DISMISSED** [review]
I'm seeing a lot of errors and state issues after this PR. But the main branch and #1198 work fine.

```
Oct 23 16:42:13.201  INFO sync:obtain_tips:download_and_verify{hash=block::Hash("00000060a9b03c2f920c3e6e0b92c5b8a253d035222dd7e6a379a126a9682135")}:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=4000 current_range=(Excluded(Height(0)), Included(Height(4000)))
Oct 23 16:42:13.601  WARN sync: zebrad::components::sync: e=Checkpoint(Duplicate { height: Height(215) })
Oct 23 16:42:13.601  INFO sync: zebrad::components::sync: waiting to restart sync timeout=45s
Oct 23 16:42:58.603  INFO sync: zebrad::components::sync: starting sync, obtaining new tips
Oct 23 16:42:58.603  INFO sync:obtain_tips: zebra_state::util: created block locator tip_height=Height(2) min_locator_height=0 locators=[Height(2), Height(1), Height(0)]
```

Note that 4000 blocks were verified, but the block locator starts at height 2.

File: zebrad/src/components/sync/downloads.rs
Comment (teor2345) [inline zebrad/src/components/sync/downloads.rs]: Do we want to error and restart the sync here?

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs:54]: I'd rather that we keep this timeout to the correct value, and just fix the hedge issue upstream -- it's not that processing the request will fail if the request takes longer than 10 seconds, but just that recording the statistics will.

File: zebrad/src/components/sync/downloads.rs
Comment (hdevalence) [reply zebrad/src/components/sync/downloads.rs]: Yes, I think that's the right behavior.

File: zebrad/src/components/sync.rs
Comment (hdevalence) [reply zebrad/src/components/sync.rs:82]: Let's merge this, so the sync works, and use it to fix the verifier pipelines.

File: zebrad/src/components/sync/downloads.rs
Comment (hdevalence) [reply zebrad/src/components/sync/downloads.rs]: Fixed in 3991cde

**Review by dconnolly — DISMISSED** [review]

**Review by dconnolly — APPROVED** [review]

**Comment by hdevalence** [comment]
Rebased again because Dependabot caused a merge conflict.  The clippy errors are unrelated to this PR.

File: zebrad/src/components/sync.rs
Comment (teor2345) [reply zebrad/src/components/sync.rs:54]: Which means that more requests are hedged than necessary, because the 95th percentile is lower. I guess that's ok.

**Pull request merged**
